### PR TITLE
`General`: Add an optional start date to exercises

### LINKED ISSUE #5803: `Communication`: Enable communication before submission time
### Is your feature request related to a problem?

_No response_

### Describe the solution you'd like

Currently, we are using two phases for our task:

Phase 1: The students get their task and are working on it themselves.
Phase 2: We enable the submission system in Artemis.

Since we are using `Discussions` this term it would be nice to have the Discussion for specific tasks before the actual submission start time.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dfuchss**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
#5803

### Description

Resolves #5803

This PR adds a "Start Date" to Exercises except Quizzes. The start date is always after the release date.
After release date, the task is visible but the students cannot start until the start date is reached.
Nevertheless, the students are still able to use communications during the period between release and start date.
Tutors and instructors can always start the exercise.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exercise (all but quiz exercise) with enabled communication features enabled

1. Log in to Artemis
2. Create an exercise in a course with enabled communication feature
3. Set Release date and Start date
4. Check that communication is possible after release date & starting is not possible before start date.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| programming-exercise-lifecycle.component.ts | 78% | 93% |

#### After Release, Before Startdate
Bildschirm­foto 2022-11-21 um 12 32 58

#### After Startdate
Bildschirm­foto 2022-11-21 um 12 39 41

#### Instructor view
Bildschirm­foto 2022-11-21 um 12 40 57

### CODE REVIEW

**Comment by dfuchss** [comment]
@JohannesStoehr I've tested one edge case:
Set no release date but the start date:
image

The start button is not locked :)

**Review by manuelmanso — APPROVED** [review]
Tested in ts1, worked as expected. Code looks good as well. I think you should add a test in the server side that tutors/instructors can still start the exercise regardless of the start date but feel free to decide.

**Comment by dfuchss** [comment]
Good idea I'll add this test tomorrow.

**Comment by dfuchss** [comment]
@JohannesStoehr yes it works :)

**Review by frederik-kt — DISMISSED** [review]
Tested locally and on TS2. Everything seems to work. Code looks good too.

**Comment by dfuchss** [comment]
@manuelmanso could you please re-approve this PR :) ?

**Comment by krusche** [comment]
The PR is not yet ready to merge: you need at least 4 approvals. I suggest to conduct a testing session in case you need those approvals urgently.

**Comment by dfuchss** [comment]
> The PR is not yet ready to merge: you need at least 4 approvals. I suggest to conduct a testing session in case you need those approvals urgently.

Thanks for the number of needed approvals. I thought it would be enough if the number of reviews and tests (check marks in the readme) are reached (that seems to be the case). Maybe I've missed some of the instructions in your dev guidelines . In this case, sorry for that.

**Review by tobias-lippert — APPROVED** [review]
Tested on ts1. Didn't find any issues. Code LGTM as well.

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java [hunk 1]
```diff
@@ -42,6 +42,10 @@ public abstract class BaseExercise extends DomainObject {
     @JsonView(QuizView.Before.class)
     private ZonedDateTime releaseDate;
 
+    @Column(name = "start_date")
+    @JsonView(QuizView.Before.class)
+    private ZonedDateTime startDate;
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: Why is this added to the `BaseExercise` entity? It seems to be just used for ProgrammingExercises right now? I would suggest moving it there. It is the same table anyway so no changes to the db layer are needed when we want to use it for other exercises in the futrue.

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 2]
```diff
@@ -789,15 +791,15 @@ public void checkCourseAndExerciseGroupExclusivity(String entityName) throws Bad
      * <p>
      * Currently, exercise start dates are the same for all users
      *
-     * @return the time from which on access to the exercise is allowed, for exercises that are not part of an exam, this is just the release date.
+     * @return the time from which on access to the exercise is allowed, for exercises that are not part of an exam, this is just the release date or start date.
      */
     @JsonIgnore
-    public ZonedDateTime getIndividualReleaseDate() {
+    public ZonedDateTime getRepositoryUnlockDate() {
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:797]: Why is this code in the `Exercise` entity. Repositories just make sense for ProgrammingExercises?

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java [hunk 3]
```diff
@@ -412,6 +412,7 @@ public void undoUnallowedChanges(QuizExercise originalQuizExercise) {
         // reset unchangeable attributes: ( dueDate, releaseDate, question.points)
         this.setDueDate(originalQuizExercise.getDueDate());
         this.setReleaseDate(originalQuizExercise.getReleaseDate());
+        this.setStartDate(originalQuizExercise.getStartDate());
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java:415]: Seems to be hack just to get start date working with QuizExercises. Like in my other comments I would suggest just using the property for programming exercises.

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Great job. Just some general questions :)

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java (hunk 1 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: This will be added to other exercises in the future

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java (hunk 3 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java:415]: Will be used for all exercises in the future

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 2 above)
Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java]: I've renamed it because it was only used at one place (and the new name fits better from my point of view, but feel free to check this). I didn't wanted to move the method in this PR.

Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java]: Reverted the renaming

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: I get that. But why not move it then to the top entity when it will be used for other exercises?

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java (hunk 3 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java:415]: See me comment from above, why do it in this PR already?

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: If you really want to leave it like this, maybe add at least a comment `Currently only used for ....` and a ToDo to introduce it to other exercises.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java (hunk 3 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizExercise.java:415]: If you really want to leave it like this, maybe add at least a comment `Currently only used for ....` and a ToDo to introduce to explain that this "hack" has to be changed once it is activated for quiz exercises.

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 4]
```diff
@@ -280,6 +280,7 @@
                 "assessmentTypeTooltipAutomaticAssessment": "Die Art der Bewertung, die nach der Einreichungsfrist durchgeführt wird: Keine manuelle Bewertung",
                 "assessmentTypeTooltipManualAssessment": "Die Art der Bewertung, die nach der Einreichungsfrist durchgeführt wird: Eine manuelle Bewertung durch Tutor:innen und Beschwerden von Studierenden, falls möglich",
                 "releaseDateTooltip": "Das Freigabedatum der Aufgabe. Studierende können erst nach dem angegebenen Datum an der Aufgabe teilnehmen.",
```
Comment (ge65cer) [inline src/main/webapp/i18n/de/programmingExercise.json:282]: Should be updated. If I understand correctly this is more like a "visible after" date now.

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 5]
```diff
@@ -280,6 +280,7 @@
                 "assessmentTypeTooltipAutomaticAssessment": "The type of assessment after the deadline: No manual assessment",
                 "assessmentTypeTooltipManualAssessment": "The type of assessment after the deadline: Manual assessment by a tutors and complaints allowed if possible",
                 "releaseDateTooltip": "The release date of the exercise. Students will only be able to participate in the exercise starting on the specified date.",
```
Comment (ge65cer) [inline src/main/webapp/i18n/en/programmingExercise.json:282]: Same as above.

**Review by ge65cer — COMMENTED** [review]
Code generally looks good, added two comments. Will test the exam mode later today.

Additionally, could you explain if and how you plan to migrate the quiz exercise's visibility date to the start date? 🙂  Do I understand correctly that this the behavior is the same? Would it maybe make sense to refactor the “visibility date” to the _BaseExercise_ class instead of introducing a new attribute?

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java (hunk 1 above)
Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: It is already working for other exercises on the server side (e.g., text exercise (server side tests)). For now it's critical for our programming exercises. Therefore, it has been implemented on the client side for that specific exercise.

**Comment by dfuchss** [comment]
> Additionally, could you explain if and how you plan to migrate the logic of the quiz exercise's visibility date (`is_visible_before_start`) to the start date? 🙂 Do I understand correctly that this the behavior is similar?

For now, this was out of scope in this PR. We need this feature for our programming tasks asap, since we need it to use the discussions feature. In the future, we might merge both values/logics to one by deleting one of the properties.

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: ok got it

**Review by stefanwaldhauser — DISMISSED** [review]
Approved since concerns were addressed and urgency.

File: src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java (hunk 1 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:47]: Added TODO for using in quiz exercise

**Review by stefanwaldhauser — APPROVED** [review]
Reapprove

**Comment by dfuchss** [comment]
@ge65cer @manuelmanso @frederik-kt @tobias-lippert 

Could you take a look at it again. @JohannesStoehr has implemented the feature fall all types of exercises (besides Quizzes)

**Review by ge65cer — DISMISSED** [review]
With release date unset and start date in future it shows “You missed the deadline”, while I would expect “You can not yet start working on this exercise”.

Screenshot 2022-11-14 at 18-00-59 Exercise details
https://artemis-test1.artemis.in.tum.de/courses/62/exercises/1107 (test user 4)

On another note, I would adapt the “You have not defined a release date for this exercise, this means that ...” alert when editing the exercise depending on if a start date was set.

**Review by laadvo — DISMISSED** [review]
Tested on ts2, created a new modeling exercise. Communication and starting the exercise worked as expected

**Review by tobias-lippert — CHANGES_REQUESTED** [review]
Tested with a text exercise on ts1.
If I set the example solution publication date to a date that is after the release date but before the start date, no error message is shown in the UI but once I want to save it, it fails, but only with a generic error message saying that some dates are not valid, I would add a client-side validation similar to the ones that already exist.

And one real edge case: if students could already start the exercise and then the instructor decides to postpone the start date but the student is still logged into Artemis and the exercise is open the student can still successfully submit something.

**Review by tobias-lippert — CHANGES_REQUESTED** [review]
Tested again on ts1. The client-side check works as expected but IMO the message should be improved.  I receive the following message if no due date is set and the example solution publication date is after the release date but before the start date but the start date is not mentioned in the error message.
image

**Review by tobias-lippert — APPROVED** [review]
All issues have been resolved. Great work.

**Review by Yunqing-Wang — COMMENTED** [review]
When I log in with my own account it says there is an internal error and no authority
image
image

**Comment by dfuchss** [comment]
> When I log in with my own account it says there is an internal error and no authority image image

What steps did you perform?

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 6]
```diff
@@ -867,15 +863,19 @@ private List<GradingInstruction> copyGradingInstruction(GradingCriterion origina
...
         if (isExamExercise()) {
             throw new BadRequestAlertException("An exam exercise may not have any dates set!", getTitle(), "invalidDatesForExamExercise");
         }
 
         // at least one is set, so we have to check the three possible errors
-        boolean areDatesValid = isNotAfterAndNotNull(getReleaseDate(), getDueDate()) && isValidAssessmentDueDate(getReleaseDate(), getDueDate(), getAssessmentDueDate())
+        boolean areDatesValid = isNotAfterAndNotNull(getReleaseDate(), getDueDate()) && isNotAfterAndNotNull(getReleaseDate(), getStartDate())
+                && isNotAfterAndNotNull(getStartDate(), getDueDate()) //
+                && isValidAssessmentDueDate(getStartDate(), getDueDate(), getAssessmentDueDate()) //
+                && isValidAssessmentDueDate(getReleaseDate(), getDueDate(), getAssessmentDueDate()) //
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:877]: I assume these comments are to stop the autoformatting? I think it would be better to use `// @formatter:off` and `// @formatter:on` here.

Also move the `isNotAfterAndNotNull(getReleaseDate(), getStartDate())` to a new line to be consistent.

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.html [hunk 7]
```diff
@@ -104,7 +104,6 @@
             [isBuilding]="false"
             [participation]="participation"
             [exercise]="exercise"
-            [short]="true"
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.html:107]: Why is this change? Or was this overlooked?

File: src/main/webapp/app/exercises/shared/course-exercises/course-exercise.service.ts [hunk 8]
```diff
@@ -158,14 +157,18 @@ export class CourseExerciseService {
 
     convertExerciseDatesFromServer<T extends Exercise>(res: T): T {
         res.releaseDate = convertDateFromServer(res.releaseDate);
+        res.startDate = convertDateFromServer(res.startDate);
         res.dueDate = convertDateFromServer(res.dueDate);
+        res.assessmentDueDate = convertDateFromServer(res.assessmentDueDate);
+        res.exampleSolutionPublicationDate = convertDateFromServer(res.exampleSolutionPublicationDate);
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/course-exercises/course-exercise.service.ts:163]: Unrelated to this PR, but I think the date of the additional test run for PE should also be converted here.

File: src/main/webapp/app/exercises/shared/exercise-headers/header-exercise-page-with-details.component.ts [hunk 9]
```diff
@@ -106,44 +108,48 @@ export class HeaderExercisePageWithDetailsComponent implements OnChanges, OnInit
...
      * Arrays (for badge class (= statusBadges) and highlighting (= isNextDueDate)) consist of
-     * 0: Submission Due Date
-     * 1: Assessment Due Date
-     * 2: Individual Complaint Deadline
-     * 3: Complaint Possible (Yes / No)
+     * 0: Start Date (either release date or start date if set)
+     * 1: Submission Due Date
+     * 2: Assessment Due Date
+     * 3: Individual Complaint Deadline
+     * 4: Complaint Possible (Yes / No)
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/exercise-headers/header-exercise-page-with-details.component.ts:131]: Can you add constants for these values and use them in the HTML and the methods?

**Review by manuelmanso — APPROVED** [review]
Tested in ts1, startdate/releasedate works as expected. There's an issue with lockign the online editor if the start date is moved to the future but it can be handled in a follow-up PR.

**Review by evgeny1984 — APPROVED** [review]
Tested in ts1, setting startdate after releasedate params works as expected.
The code changes are also fine.

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 6 above)
Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:878]: I agree. I'll look at it later

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.html (hunk 7 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.html:107]: Ah, seams I committed this in the wrong pr. But I wanted to change this non the less.
Do you think this causes any problems here?

File: src/main/webapp/app/exercises/shared/course-exercises/course-exercise.service.ts (hunk 8 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/course-exercises/course-exercise.service.ts:163]: This method handles all exercise types, so we cannot do this here

**Comment by JohannesStoehr** [comment]
Last commits only changed styling, manual reviews still valid

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 10]
```diff
@@ -789,15 +785,15 @@ public void checkCourseAndExerciseGroupExclusivity(String entityName) throws Bad
      * <p>
      * Currently, exercise start dates are the same for all users
      *
-     * @return the time from which on access to the exercise is allowed, for exercises that are not part of an exam, this is just the release date.
+     * @return the time from which on access to the exercise is allowed, for exercises that are not part of an exam, this is just the release date or start date.
      */
     @JsonIgnore
     public ZonedDateTime getIndividualReleaseDate() {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:791]: I think the name is misleading now because "release" has a dedicated meaning

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 11]
```diff
@@ -155,16 +156,23 @@ public ResponseEntity<Participation> startParticipation(@PathVariable Long exerc
         log.debug("REST request to start Exercise : {}", exerciseId);
         Exercise exercise = exerciseRepository.findByIdElseThrow(exerciseId);
         User user = userRepository.getUserWithGroupsAndAuthorities();
-        Participant participant = user;
         authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.STUDENT, exercise, user);
 
         // if the user is a student and the exercise has a release date, they cannot start the exercise before the release date
-        if (exercise.getReleaseDate() != null && exercise.getReleaseDate().isAfter(now())) {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:162]: I don't see the reason behind this change as you still compare the start date afterwards.

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts [hunk 12]
```diff
@@ -94,10 +94,34 @@ export class ProgrammingExerciseLifecycleComponent implements OnInit, OnChanges
             // Changes from parent component are allowed but no cascading changes should be made in read-only mode.
             return;
         }
+        if (this.exerciseService.hasStartDateError(this.exercise)) {
+            this.updateStartDate(newReleaseDate);
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts:98]: I realise this has been done before but do we rally want to change some other date on error? I think it's counter-intuitive as a user that when an error is shown, stuff changes automatically.

File: src/main/webapp/app/exercises/shared/exercise/exercise.service.ts [hunk 13]
```diff
@@ -72,14 +72,20 @@ export class ExerciseService {
...
     validateDate(exercise: Exercise) {
         exercise.dueDateError = this.hasDueDateError(exercise);
+        exercise.startDateError = this.hasStartDateError(exercise);
         exercise.assessmentDueDateError = this.hasAssessmentDueDateError(exercise);
 
         exercise.exampleSolutionPublicationDateError = this.hasExampleSolutionPublicationDateError(exercise);
         exercise.exampleSolutionPublicationDateWarning = this.hasExampleSolutionPublicationDateWarning(exercise);
     }
 
+    hasStartDateError(exercise: Exercise) {
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:82]:
```suggestion
    isInvalidStartDate(exercise: Exercise) {
```
I think that phrasing makes more sense, doesn't it?

File: src/main/webapp/app/exercises/shared/exercise/exercise.service.ts [hunk 14]
```diff
@@ -72,14 +72,20 @@ export class ExerciseService {
...
+        return exercise.startDate && exercise.releaseDate && dayjs(exercise.startDate).isBefore(exercise.releaseDate);
+    }
+
     hasDueDateError(exercise: Exercise) {
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:86]:
```suggestion
    isInvalidDueDate(exercise: Exercise) {
```
I think that phrasing makes more sense, doesn't it?

Comment (julian-christl) [reply src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:86]: If yes, please fix similar occurrences in this service

File: src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts [hunk 15]
```diff
@@ -196,10 +198,22 @@ export const participationStatus = (exercise: Exercise, testRun?: boolean): Part
...
 
-export const isStartExerciseAvailable = (exercise: ProgrammingExercise): boolean => {
-    return !exercise.dueDate || dayjs().isBefore(exercise.dueDate);
+/**
+ * Determines if the exercise can be started, this is the case if:
+ * - It is after the start date or the participant is at least a tutor
+ * - In case of a programming exercise it is not before the due date
+ * @param exercise the exercise that should be started
+ */
+export const isStartExerciseAvailable = (exercise: Exercise): boolean => {
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts:207]:
```suggestion
export const canStartExercise = (exercise: Exercise): boolean => {
```

```suggestion
export const allowExerciseStart = (exercise: Exercise): boolean => {
```

With the current wording, I would expect a check whether something is present in the entity

Comment (julian-christl) [reply src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts:207]: If you agree, please fix similar occurrences in this service

File: src/main/webapp/i18n/de/exercise.json [hunk 16]
```diff
@@ -74,13 +74,15 @@
             "title": "Titel",
             "shortName": "Kurzname",
             "releaseDate": "Veröffentlichungsdatum",
+            "startDate": "Startdatum",
             "dueDate": "Einreichungsfrist",
-            "dueDateError": "Einreichungsfrist muss nach dem Veröffentlichungsdatum sein. Bitte überprüfe deine Auswahl nochmal",
+            "dueDateError": "Einreichungsfrist muss nach dem Veröffentlichungsdatum und falls vorhanden dem Startdatum sein. Bitte überprüfe deine Auswahl nochmal.",
```
Comment (julian-christl) [inline src/main/webapp/i18n/de/exercise.json:79]:
```suggestion
            "dueDateError": "Einreichungsfrist muss nach dem Veröffentlichungsdatum und, falls vorhanden, dem Startdatum sein. Bitte überprüfe deine Auswahl nochmal.",
```

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 17]
```diff
@@ -279,7 +280,8 @@
                 "automaticTestsAfterDueDateTooltip": "Sobald die Einreichungsfrist abgelaufen ist, wird der Build für alle Abgaben erneut durchgeführt und sie werden ein letztes Mal automatisch getestet unter Verwendung ALLER im Test-Repository gespeicherten Tests.",
                 "assessmentTypeTooltipAutomaticAssessment": "Die Art der Bewertung, die nach der Einreichungsfrist durchgeführt wird: Keine manuelle Bewertung",
                 "assessmentTypeTooltipManualAssessment": "Die Art der Bewertung, die nach der Einreichungsfrist durchgeführt wird: Eine manuelle Bewertung durch Tutor:innen und Beschwerden von Studierenden, falls möglich",
-                "releaseDateTooltip": "Das Freigabedatum der Aufgabe. Studierende können erst nach dem angegebenen Datum an der Aufgabe teilnehmen.",
+                "releaseDateTooltip": "Das Freigabedatum der Aufgabe. Studierende können die Aufgabe ab diesem Datum sehen. Sollte kein Startdatum festgelegt sein können Studenten auch ab diesem Datum die Aufgabe beginnen.",
```
Comment (julian-christl) [inline src/main/webapp/i18n/de/programmingExercise.json:283]:
```suggestion
                "releaseDateTooltip": "Das Freigabedatum der Aufgabe. Studierende können die Aufgabe ab diesem Datum sehen. Sollte kein Startdatum festgelegt sein können Studierdende auch ab diesem Datum die Aufgabe beginnen.",
```

File: src/main/webapp/i18n/de/student-dashboard.json [hunk 18]
```diff
@@ -190,6 +190,8 @@
                 "rated": "Bewertet",
                 "practice": "Übung",
                 "assessmentType": "Bewertung:",
+                "startDate": "Startdatum:",
+                "startDateTooltip": "Student:innen können diese Aufgabe ab {{date}} bearbeiten.",
```
Comment (julian-christl) [inline src/main/webapp/i18n/de/student-dashboard.json:194]:
```suggestion
                "startDateTooltip": "Studierende können diese Aufgabe ab {{date}} bearbeiten.",
```
According to our guidelines.

File: src/test/javascript/spec/component/exercises/shared/headers/header-exercise-page-with-details.component.spec.ts [hunk 19]
```diff
@@ -66,8 +66,8 @@ describe('HeaderExercisePageWithDetails', () => {
         component.ngOnInit();
 
         expect(component.exerciseCategories).toEqual([]);
-        expect(component.isNextDueDate).toStrictEqual([false, false, false, false]);
-        expect(component.statusBadges).toStrictEqual(['bg-danger', 'bg-danger', 'bg-danger']);
+        expect(component.nextDueDateIndex).toBeUndefined();
```
Comment (julian-christl) [inline src/test/javascript/spec/component/exercises/shared/headers/header-exercise-page-with-details.component.spec.ts:69]: You probably want to check these action errors

**Review by julian-christl — CHANGES_REQUESTED** [review]

File: src/test/javascript/spec/component/exercises/shared/headers/header-exercise-page-with-details.component.spec.ts (hunk 19 above)
Comment (julian-christl) [reply src/test/javascript/spec/component/exercises/shared/headers/header-exercise-page-with-details.component.spec.ts:69]: Client tests are failing too, probably due to this.

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts (hunk 12 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts:98]: This is the current behavior of the complete component.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java (hunk 11 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:162]: getIndividualReleaseDate also takes the StartDate into account

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.html (hunk 7 above)
Comment (Strohgelaender) [reply src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.html:107]: no, just wanted to ensure this is an intentional change.

**Review by Strohgelaender — APPROVED** [review]
Code changes since last review look good.

**Review by manuelmanso — APPROVED** [review]
Last changes LGTM

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java (hunk 11 above)
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:162]: Exactly. If it gets taken into account, why do we need to check the startDate afterwards?

File: src/main/webapp/i18n/de/assessment.json [hunk 20]
```diff
@@ -64,7 +64,7 @@
                 "warning": "Achtung: Die Einreichungsfrist ist noch nicht abgelaufen und Studierende dürfen noch ihre Einreichungen abgeben. Bitte noch nicht mit der Korrektur anfangen.",
                 "warningIndividual": "Achtung: Die allgemeine Einreichungsfrist ist abgelaufen, aber manche Studierende mit verlängerter Bearbeitungszeit dürfen noch ihre Einreichungen abgeben. Bitte stelle vor Beginn einer Korrektur sicher, dass diese Studierenden nicht mehr abgeben können.",
                 "empty": "Keine Abgabe gefunden!",
-                "emptySearch": "Bitte gib einen Studentennamen ein, um nach Abgaben zu suchen.",
+                "emptySearch": "Bitte gib einen Studierendennamen ein, um nach Abgaben zu suchen.",
```
Comment (julian-christl) [inline src/main/webapp/i18n/de/assessment.json:67]:
```suggestion
                "emptySearch": "Bitte gib den Namen eines Studierenden ein, um nach Abgaben zu suchen.",
```
Weird neologism/compound word imo

**Review by julian-christl — APPROVED** [review]
Required changes included. Thank you.

Code approved.

Please check once develop is fixed that tests and build run through again.

**Review by evgeny1984 — DISMISSED** [review]
The code changes look good to me since last change request.
Green light!

**Review by manuelmanso — DISMISSED** [review]
Code still looks good with latest changes

**Review by frederik-kt — DISMISSED** [review]
Code looks good!

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove! Good Job

**Comment by dfuchss** [comment]
@ge65cer can you also reapprove it? Afterwards, github should be happy :)

**Review by Yunqing-Wang — COMMENTED** [review]
I still have this error message
image
And it is possible to post a communication message after the release but before starting date. However, wenn I start the exercise before the state date, I got this:
image

**Comment by dfuchss** [comment]
@JohannesStoehr do you have an idea?

**Review by laadvo — COMMENTED** [review]
The realese date and start date seem to be working fine , but when I exit a an exercise I also get an Internal server error quite often.

**Comment by krusche** [comment]
> The realese date and start date seem to be working fine , but when I exit a an exercise I also get an Internal server error quite often.

Can you please add more details, it is difficult to reproduce the issue otherwise

**Comment by laadvo** [comment]
> > The realese date and start date seem to be working fine , but when I exit a an exercise I also get an Internal server error quite often.
> 
> Can you please add more details, it is difficult to reproduce the issue otherwise

I created an exercise with realese and start date set in a few minutes. I took part in the exercise after the start date and it worked, but then went I went back a page (to the exercises overview) I got an internal server error. This also happened when I clicked on the same exercise again and then left it.

**Comment by JohannesStoehr** [comment]
> @JohannesStoehr do you have an idea? Is this related to our PR?

Probably unrelated. The repository clone button is not touched at all

**Comment by JohannesStoehr** [comment]
> The realese date and start date seem to be working fine , but when I exit a an exercise I also get an Internal server error quite often.

ts1 had some of these lately. Did you check it in another course as well or on another test instance (locally or another server)?

**Review by laadvo — APPROVED** [review]
It seems to be a ts1 related problem, no internal errors when testing on ts2.

**Comment by dfuchss** [comment]
Just merged master.xml. Reviews should still be valid. (Please only verify that I correctly merged master.xml :) )

**Review by Yunqing-Wang — COMMENTED** [review]
still get this
image
when posting a comment before starting date
image
And not able to clone
image

**Comment by dfuchss** [comment]
Where did you tested it? There seems to be a problem on ts1. 

> ts1 had some of these lately. Did you check it in another course as well or on another test instance (locally or another server)?

**Review by julian-christl — DISMISSED** [review]
Code Reapproval

**Comment by krusche** [comment]
> still get this
> image
> when posting a comment before starting date
> image
> And not able to clone
> image

I think I resolved the issue on TS1: some data in the database was incorrect (due to old branches being used) and led to the issues.

Please note: It should explicitly be possible to post questions before the start date of the exercise!
The clone dialog should not be distorted however

**Comment by dfuchss** [comment]
@Yunqing-Wang does this resolve the issue

> > still get this
> > image
> > when posting a comment before starting date
> > image
> > And not able to clone
> > image
> 
> I think I resolved the issue on TS1: some data in the database was incorrect (due to old branches being used) and led to the issues.
> 
> Please note: It should explicitly be possible to post questions before the start date of the exercise! The clone dialog should not be distorted however

**Review by Yunqing-Wang — COMMENTED** [review]
new problem on ts1: not able to reset the start and release date
image

**Comment by dfuchss** [comment]
> new problem on ts1: not able to reset the start and release date
> image
> 

Mm maybe test it on another test server :)

**Review by laadvo — APPROVED** [review]
Reapprove for ts2

**Review by evgeny1984 — CHANGES_REQUESTED** [review]
Everything works good, tested on ts2.
I miss the information about the start date on the exercise overview screen, if it was set.
image

**Review by evgeny1984 — APPROVED** [review]
Everything works as expected on ts2. Thank you for integrating my feedback!

**Review by timo07k — CHANGES_REQUESTED** [review]
I could start the exercise although the Start Date is still in a day. Tested on TS1.

**Comment by JohannesStoehr** [comment]
> I could start the exercise although the Start Date is still in a day. Tested on TS1.

@timo07k You were probably logged in as a tutor or instructor. Pleas try again using an account that is only a student

**Review by manuelmanso — APPROVED** [review]
Tested in ts1 and it's working, the exercise can only be started after start date. Last changes also look good.

**Comment by dfuchss** [comment]
@timo07k do you have more input for us. Your request is currently the only one that blocks that PR :)

**Review by laadvo — APPROVED** [review]
Reapproved, tested on ts2
